Add compression support in JdkClientHttpRequestFactory

spicydev opened the pull request:
### Title
Implement custom BodyHandler for GZIP decompression in JdkClientHttpRequest

### Description

**Problem:** JDK HTTP Client doesn't automatically decompress responses.

**Solution:** creating a custom BodyHandler that checks Content-Encoding and applies GZIPInputStream when necessary.
**Key Changes:** 
- Added new boolean field `compressionEnabled` to `JdkClientHttpRequestFactory` and `JdkClientHttpRequest`.
- Modified `buildRequest()` method to add `Accept-Encoding` header with `gzip` value when compression is enabled.
- Added a custom `DecompressingBodyHandler` that implements `BodyHandler<InputStream>`. The apply() method in the custom handler checks the `Content-Encoding` header.
- If the encoding is `gzip`, it uses BodySubscribers.mapping to wrap the InputStream with GZIPInputStream.
  Otherwise, it uses the default BodySubscribers.ofInputStream().

**Testing Steps:**
- Make an HTTP request to an endpoint that returns a gzip-encoded response (e.g., https://vpic.nhtsa.dot.gov/api/vehicles/getallmanufacturers?format=json).
- Verify that the response body is correctly decompressed and the data is accessible.

spicydev committed:
Adds gzip handler to JDK http client based on header

Added new flag to JdkClientHttpRequest and created new DecompressingBodyHandler class to handle HttpClient response body decompression when encoded. Closes gh-35222

Signed-off-by: spicydev <vivek@mirchi.dev>

bclozel commented:
Should we also support the "deflate" compression algorithm?

Also, I am wondering if we really need an option that enables compression. Should we enable this no matter what?

What do you think?

spicydev commented:
Hi @bclozel, First off let me start off by apologizing for raising the PR in such bad state. I’ll work on unit tests and follow style guide once we can finalize on the requirements.

coming to your questions, Yes we can add support for multiple encoding types but following other clients like apache httpclient & netty, they only support gzip by default out of the box. I think at-least for now we can add support for gzip to JDK http client.

I think it’s better to have the control to the users but by default we can enable compression as it causes no harm but improves performance for all users.

Let me know your thoughts, based on that i can go ahead with changes.

bclozel commented:
@spicydev no worries, we are here to help.

I think Apache HttpClient [supports both](https://github.com/apache/httpcomponents-client/tree/9d38e5fb4438e8fb6442ac8e5ff304142938cf83/httpclient5/src/main/java/org/apache/hc/client5/http/entity). And [so does Netty](https://github.com/netty/netty/blob/4.2/codec-http/src/main/java/io/netty/handler/codec/http/HttpContentDecompressor.java).

spicydev committed:
Added support for deflate encoding and test cases

Added support for deflate encoded response using InflaterInputStream and added Unit Test cases.

Signed-off-by: spicydev <vivek@mirchi.dev>

spicydev commented:
Hi @bclozel , Need your help here. When testing the Inflater with JDK http client I'm seeing error like below
`java.util.zip.ZipException: incorrect header check`. Can you check the unit test once

bclozel commented:
Thanks a lot @spicydev for your first contribution! Tests were initially failing because the test server was writing non-printable characters as a String instead of bytes directly.

@driverpt I don't think we should open the API more here. A `BodyHandler` is very much a JDK client concern. If a custom body handler is needed, developers should use the `HttpClient` directly.

driverpt commented:
Will this be backported into current Spring Boot 3.5.x?

bclozel commented:
We are all good now I think @spicydev
